source/agora/consensus/EnrollmentManager.d

```diff
@@ -682,6 +639,56 @@ public class EnrollmentManager
         }
     }
 
+    /***************************************************************************
+
+        Generate pre-images
+
+        Params:
+            current_height = block height used to determine which range of
+                pre-images will be generated.
+
+    ***************************************************************************/
+
+    private void generatePreimages(ulong current_height) @safe nothrow
+    {
+        // if there is no pre-image, the defined number of pre-images must
+        // be generated with the interval of ValidatorCycle.
+        if (this.preimages.length == 0)
+        {
+            auto hash = hashFull(node_key_pair.secret);
+            uint current_idx = AllCountPrimages;
+            this.preimages[current_idx] = hash;
+            foreach (ulong i; 0 .. AllCountPrimages)
+            {
+                hash = hashFull(hash);
+                if (--current_idx % ValidatorSet.ValidatorCycle == 0)
+                    this.preimages[current_idx] = hash;
+            }
+        }
+
+        // This determines which range of preimages must be generated.
+        ulong start_index =
+            ((current_height / ValidatorSet.ValidatorCycle) + 2) *
+                ValidatorSet.ValidatorCycle;
+
+        // This generates all the sequential pre-images from the start index
+        auto hash = this.preimages[start_index];
+        foreach (int i; 1 .. ValidatorSet.ValidatorCycle * 2)
+        {
+            hash = hashFull(hash);
+            this.preimages[--start_index] = hash;
+        }
+
+        // Old pre-images are removed to save memory
+        () @trusted {
+            foreach (ulong i; this.preimages.keys)
```

Comment:
This allocates a new array that will store all the keys.

There is an alternative `preimages.byKey`, however it cannot be used here. This is because doing something like this:

```D
foreach (key; this.preimages.byKey)
    this.preimages.remove(key);
```

As soon as `remove` is called, the `byKey` range is in an invalid state. So the above code should not be used.

I have suggested an alternative in the main PR comment section.

Reply:
> There is an alternative preimages.byKey, however it cannot be used here. 

If `cur_cycle_preimages` is applied to the code, the code removing old pre-images is not needed anymore.